nntrainer/layers/mol_attention_layer.cpp

```diff
@@ -116,19 +202,159 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
+  Tensor beta_eps = beta.add(1e-8f);
+
+  Tensor u_pos_m = u_pos.subtract(m);
+  u_pos_m.divide(beta_eps, u_pos_div);
+  sigmoid.run_fn(u_pos_div, prob_left);
 
-  Tensor integral_scaled = integral.multiply(alpha);
-  Tensor scores = integral_scaled.sum(3);
-  scores.reshape(TensorDim({scores.batch(), 1, 1, scores.height()}));
+  Tensor u_neg_m = u_neg.subtract(m);
+  u_neg_m.divide(beta_eps, u_neg_div);
+  sigmoid.run_fn(u_neg_div, prob_right);
+
+  prob_left.subtract(prob_right, prob);
+
+  Tensor prob_scaled = prob.multiply(alpha);
+  prob_scaled.sum(3, scores);
 
   scores.dotBatched(value, output);
 }
 
-void MoLAttentionLayer::calcDerivative(RunLayerContext &context) { /** NYI */
+void MoLAttentionLayer::calcDerivativeHelper(RunLayerContext &context,
+                                             Tensor &dstate) {
+  Tensor &query = context.getInput(wt_idx[AttentionParams::query]);
+  Tensor &value = context.getInput(wt_idx[AttentionParams::value]);
+
+  Tensor &derivative = context.getIncomingDerivative(SINGLE_INOUT_IDX);
+
+  Tensor &fc_proj_out = context.getTensor(wt_idx[AttentionParams::fc_proj_out]);
+  Tensor &dfc_proj_out =
+    context.getTensor(wt_idx[AttentionParams::fc_proj_out]);
+  Tensor &scores = context.getTensor(wt_idx[AttentionParams::scores]);
+  Tensor &prob = context.getTensor(wt_idx[AttentionParams::prob]);
+  Tensor &prob_left = context.getTensor(wt_idx[AttentionParams::prob_left]);
+  Tensor &prob_right = context.getTensor(wt_idx[AttentionParams::prob_right]);
+  Tensor &u_neg_div = context.getTensor(wt_idx[AttentionParams::u_neg_div]);
+  Tensor &u_pos_div = context.getTensor(wt_idx[AttentionParams::u_pos_div]);
+
+  const TensorDim &input_dim = query.getDim();
+  unsigned int batch = input_dim.batch();
+  auto mol_k = std::get<props::MoL_K>(mol_props).get();
+
+  Tensor kappa, beta, alpha;
+  kappa.copy_with_stride(
+    fc_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, 0, false));
+  beta.copy_with_stride(
+    fc_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, mol_k, false));
+  alpha.copy_with_stride(
+    fc_proj_out.getSharedDataTensor({batch, 1, 1, mol_k}, mol_k * 2, false));
+
+  Tensor dscores = Tensor(TensorDim({value.batch(), 1, 1, value.height()}));
+  dscores.dot_batched_deriv_wrt_1(value, derivative);
+  dscores.reshape(TensorDim({scores.batch(), 1, scores.width(), 1}));
+
+  Tensor dprob_scaled = Tensor(TensorDim({batch, 1, value.height(), mol_k}));
+  dprob_scaled.setZero();
+  dprob_scaled.add_i(dscores);
+
+  Tensor dalpha = dprob_scaled.multiply(prob).sum(2);
+  Tensor dprob = dprob_scaled.multiply(alpha);
+
+  Tensor dprob_left = dprob;
+  Tensor dprob_right = dprob.multiply(-1);
+
+  Tensor beta_eps = beta.add(1e-8f);
+  Tensor du_neg_div, du_pos_div;
+  sigmoid.run_prime_fn(prob_right, du_neg_div, dprob_right);
+  Tensor du_neg_m = du_neg_div.divide(beta_eps);
+  Tensor dm_neg = du_neg_m.multiply(-1).sum(2);
+  Tensor dbeta_eps_neg = du_neg_m.multiply(u_neg_div).multiply(-1).sum(2);
```

Comment:
considering) u_neg_div = u_neg_m/beta_eps

dbeta_eps = -d_u_neg_div * u_neg_m / (beta_eps ^ 2) ?

Reply:
your formula is right. To reduce the amount of computation, consider the below subsitutions and simplified result used in the code:
```
du_neg_m = d_u_neg_div/beta_eps // given
u_neg_div = u_neg_m/beta_eps // given
dbeta_eps/du_neg_div = -d_u_neg_div * u_neg_m / (beta_eps ^ 2) = (d_u_neg_div/beta_eps) * (u_neg_m/beta_eps) * -1 = du_neg_m * u_neg_div
```

Reply:
That's clever. Thanks for the explanation :)